modules/renter/upload.go

```diff
@@ -62,47 +62,46 @@ func (r *Renter) threadedUploadPiece(up modules.FileUploadParams, piece *filePie
 		// Select a host. An error here is unrecoverable.
 		host, err := r.hostDB.RandomHost()
 		if err != nil {
-			return
+			return err
 		}
 
 		// Negotiate the contract with the host. If the negotiation is
-		// unsuccessful, we need to try again with a new host. Otherwise, the
-		// file will be uploaded and we'll be done.
+		// unsuccessful, we need to try again with a new host.
 		contract, contractID, key, err := r.negotiateContract(host, up)
-		if err != nil {
-			// The previous attempt didn't work. We will try again after
-			// sleeping for a randomized amount of time to increase our chances
-			// of success. This will help spread things out if there are
-			// problems with network congestion or other randomized issues.
-			randSource := make([]byte, 1)
-			rand.Read(randSource)
-			time.Sleep(time.Duration(attempts) * time.Duration(attempts) * 250 * time.Millisecond * time.Duration(randSource[0]))
-			continue
+		if err == nil {
+			// Negotiation was successful; update the filePiece.
+			lockID := r.mu.Lock()
+			*piece = filePiece{
+				Active:     true,
+				Repairing:  false,
+				Contract:   contract,
+				ContractID: contractID,
+
+				HostIP: host.IPAddress,
+
+				EncryptionKey: key,
+			}
+			r.save()
+			r.mu.Unlock(lockID)
+			return nil
 		}
 
-		lockID := r.mu.Lock()
-		*piece = filePiece{
-			Active:     true,
-			Repairing:  false,
-			Contract:   contract,
-			ContractID: contractID,
-
-			HostIP: host.IPAddress,
-
-			EncryptionKey: key,
-		}
-		r.save()
-		r.mu.Unlock(lockID)
-		return
+		// The previous attempt didn't work. We will try again after
+		// sleeping for a randomized amount of time to increase our chances
+		// of success. This will help spread things out if there are
+		// problems with network congestion or other randomized issues.
+		randSource := make([]byte, 1)
+		rand.Read(randSource)
+		time.Sleep(time.Duration(attempts) * time.Duration(attempts) * 250 * time.Millisecond * time.Duration(randSource[0]))
 	}
+
+	// All attempts failed.
+	return errors.New("failed to upload filePiece")
```

Comment:
yeah that shouldn't be too hard since you have the whole set of errors available. stick them in a `map[error]int` or something to pick the most frequent one.